packages/tldraw/src/lib/shapes/shared/defaultStyleDefs.tsx

```diff
@@ -154,86 +147,111 @@ const canvasBlob = (size: [number, number], fn: (ctx: CanvasRenderingContext2D)
 	fn(ctx)
 	return canvas.toDataURL()
 }
-type PatternDef = { zoom: number; url: string; darkMode: boolean }
-
-const getDefaultPatterns = () => {
-	const defaultPatterns: PatternDef[] = []
-	for (let i = 1; i <= HASH_PATTERN_COUNT; i++) {
-		const whitePixelBlob = canvasBlob([1, 1], (ctx) => {
-			ctx.fillStyle = DefaultColorThemePalette.lightMode.black.semi
-			ctx.fillRect(0, 0, 1, 1)
-		})
-		const blackPixelBlob = canvasBlob([1, 1], (ctx) => {
-			ctx.fillStyle = DefaultColorThemePalette.darkMode.black.semi
-			ctx.fillRect(0, 0, 1, 1)
-		})
-		defaultPatterns.push({
-			zoom: i,
-			url: whitePixelBlob,
-			darkMode: false,
-		})
-		defaultPatterns.push({
-			zoom: i,
-			url: blackPixelBlob,
-			darkMode: true,
-		})
+type PatternDef = { zoom: number; url: string; theme: 'light' | 'dark' }
+
+let defaultPixels: { white: string; black: string } | null = null
+function getDefaultPixels() {
+	if (!defaultPixels) {
+		defaultPixels = {
+			white: canvasBlob([1, 1], (ctx) => {
+				ctx.fillStyle = '#f8f9fa'
+				ctx.fillRect(0, 0, 1, 1)
+			}),
+			black: canvasBlob([1, 1], (ctx) => {
+				ctx.fillStyle = '#212529'
+				ctx.fillRect(0, 0, 1, 1)
+			}),
+		}
+	}
+	return defaultPixels
+}
+
+function getPatternLodForZoomLevel(zoom: number) {
+	return Math.ceil(Math.log2(Math.max(1, zoom)))
+}
+
+export function getHashPatternZoomName(zoom: number, theme: TLDefaultColorTheme['id']) {
+	const lod = getPatternLodForZoomLevel(zoom)
+	return `tldraw_hash_pattern_${theme}_${lod}`
+}
+
+function getPatternLodsToGenerate(maxZoom: number) {
+	const levels = []
+	const minLod = 0
+	const maxLod = getPatternLodForZoomLevel(maxZoom)
+	for (let i = minLod; i <= maxLod; i++) {
+		levels.push(Math.pow(2, i))
 	}
-	return defaultPatterns
+	return levels
+}
+
+function getDefaultPatterns(maxZoom: number): PatternDef[] {
+	const defaultPixels = getDefaultPixels()
+	return getPatternLodsToGenerate(maxZoom).flatMap((zoom) => [
+		{ zoom, url: defaultPixels.white, theme: 'light' },
+		{ zoom, url: defaultPixels.black, theme: 'dark' },
+	])
 }
 
 function usePattern() {
 	const editor = useEditor()
-	const dpr = editor.getInstanceState().devicePixelRatio
+	const dpr = useValue('devicePixelRatio', () => editor.getInstanceState().devicePixelRatio, [
+		editor,
+	])
+	const maxZoom = useValue('maxZoom', () => Math.ceil(last(editor.getCameraOptions().zoomSteps)!), [
```

Comment:
@steveruizok is this the best way to get the max zoom level?

Reply:
this seems right to me, let's go with it for now and @steveruizok can chime in later if it's wrong